crengine/src/lvdocview.cpp

```diff
@@ -3542,8 +3748,21 @@ static void FileToArcProps(CRPropRef props) {
 	props->setHex(DOC_PROP_FILE_CRC32, 0);
 }
 
+static bool needToConvertBookmarks(CRFileHistRecord* historyRecord)
+{
+    bool convertBookmarks = false;
+    if(historyRecord) {
+        gDOMVersionRequested = historyRecord->getDOMversion();
+        if(gDOMVersionRequested < 20180528) {
+            convertBookmarks = historyRecord->getBookmarks().length() > 1;
+        }
+    } else
+        gDOMVersionRequested = gDOMVersionCurrent;
```

Comment:
Been wondering how much of that CRFileHistRecord stuff in that LVDocView::LoadDocument() I can keep without risk for KOReader, as we don't use CRFileHistRecord, so all that stuff should be no-op.
Is that the case too with CR on Android?
But this line bothers me: `needToConvertBookmarks()` is called even if there's no CRFileHistRecord (meaning all that is managed from outside), so it looks like this `else` is wrongly killing the gDOMVersionRequested the outside code may have requested?

Reply:
If we change needToConvertBookmarks() like below:

```
diff --git a/crengine/src/lvdocview.cpp b/crengine/src/lvdocview.cpp
index d2ea6dc..ff7928e 100644
--- a/crengine/src/lvdocview.cpp
+++ b/crengine/src/lvdocview.cpp
@@ -3751,13 +3751,12 @@ static void FileToArcProps(CRPropRef props) {
 static bool needToConvertBookmarks(CRFileHistRecord* historyRecord)
 {
     bool convertBookmarks = false;
-    if(historyRecord) {
+    if(historyRecord && historyRecord->getBookmarks().length() > 1) {
         gDOMVersionRequested = historyRecord->getDOMversion();
         if(gDOMVersionRequested < 20180528) {
-            convertBookmarks = historyRecord->getBookmarks().length() > 1;
+            convertBookmarks = true;
         }
-    } else
-        gDOMVersionRequested = gDOMVersionCurrent;
+    }
     return convertBookmarks;
 }
```

This way it will not affect you, right?

Reply:
That looks fine. (So, that way, you're not resetting at all gDOMVersionRequested , but it will be gDOMVersionCurrent when you have no historyRecord because... you set/reset it to gDOMVersionCurrent elsewhere?)

Reply:
Initially gDOMVersionRequested is equal to gDOMVersionCurrent. I did not find were is it changed. I assume it is changed in a frontend or I missed something. So the one who changes gDOMVersionRequested  is responsible for resetting it.

I did some more changes, please take a look: https://gist.github.com/pkb/64714978e3a467b687b738c3a0f3b51b

Does it make sense?

Reply:
> So the one who changes gDOMVersionRequested is responsible for resetting it.

OK, In KOReader, we set it from frontend before each book loading (from the value saved in our book settings, or to the gDOMVersionCurrent that we have fetched from CRE, when the book has no setting yet).
I get that in your case, for each book, you set it first to the value saved in your CRHistFile (so, the oldest if none), always migrate bookmarks and always end up setting it to gDOMVersionCurrent. So I guess both you and me are fine (as long as you don't tweak gDOMVersionRequested when there is no CRHistFile :)
Your patch looks sane to me, when reading it in your and my contexts.

But I don't know how that work for CR Android, You told me bookmarks were saved in a SQLite DB, so I assume it's not using CRHistFile, right?. Is that gDOMVersion business and bookmarks migration taken care in any way on CR Android?

Also, a question, just because I don't know much about CRHistFile: is it always fed from outside frontends? So, if that is not done by the frontend, it's always NULL and checks with it are just skipped, right?

Btw, in your merged code and patch, you might want to use (instead of your hardcoded `< 20180528`):
```c
#define DOM_VERSION_WITH_XPOINTERS_V2 20180528
or
#define DOM_VERSION_WITH_NORMALIZED_XPOINTERS 20180528
```
(because for me, that won't be `20180528`, but something like `20200214`, depends on when I'll get that ready :) or you could have used anything 2020xyxz that I could reuse - but minor diffs, no problem).

`historyRecord->getBookmarks().length() > 1` - another (less critical) thingy that might need conversion is the last position in book. After migration and switch to the latest DOMVersion, the last position might not resolve - unless crengine handles that somehow (I've seen it doing that in some other contexts, like re-renderings, but not sure if it uses a xpointer or a y-coordinate).

(In CRFileHistRecord::convertBookmarks(), your toggling of gDOMVersionRequested  between old and new versions is just to pick the right version of createXPointer() and toString(), right? would have worked the same by using createXPointerV1 and toStringUsingNames(New), right?)

Reply:
I don't know much about Android part. Let's wait and see what @virxkane has to say about it.
AFAIU new CRFileHistRecord gets created whenever LVDocView::savePosition() is called. So you might have a record if you call LVDocView::goLink() for example. It will not be saved to a file, so on start record list will be empty. But anyway getBookmarks().length() will be at most 1 so 

> Btw, in your merged code and patch, you might want to use

Ok

> another (less critical) thingy that might need conversion is the last position in book

Yes I'm aware of this. Maybe I'm wrong but I've decided to skip this conversion. There is a chance that reading position will be lost.

> In CRFileHistRecord::convertBookmarks(), your toggling of gDOMVersionRequested

Right.

@virxkane could you please check the patch https://gist.github.com/pkb/64714978e3a467b687b738c3a0f3b51b and apply it along with your other changes if you find it ok? Thanks

Reply:
@pkb Ok, it's done. https://github.com/virxkane/coolreader/commit/999a484ee1f9c4a9287932537fae9697f96f7d28
@poire-z About your question: yes, on Android bookmarks saved in SQLite DB, so CRHistFile is unused. Bookmark migration on Android for now not implemented yet. In latest commit https://github.com/virxkane/coolreader/commit/3189393b1067ddab0b33375ebbf1f40014258edb I implement saving & loading DOM version of each document in DB, so at least old bookmark must be handled properly. In the future, the conversion of bookmarks will probably be implemented.

Reply:
Some update from my side: I brought these normalize xpointers in https://github.com/koreader/crengine/pull/328 - with some additional minor tweaks and renames.

> AFAIU new CRFileHistRecord gets created whenever LVDocView::savePosition() is called. So you might have a record if you call LVDocView::goLink() for example.

Ok, we're indeed using that, only when resizing or changing orientations. Didn't take any risk and let your CRHistFile conversion code out.